### PR TITLE
[MASTER] Release/v1.0.0

### PR DESCRIPTION
**Pull request by zackharley**
Initial release!

### CODE REVIEW

File: client/ClientStore.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import {
+	rootSaga as hackerSaga,
+	reducer as hackerReducer,
+	reducerMount as HACKER_MOUNT
+} from './HackerStore';
+
+// reducer mount for redux form reducer
+const FORM_MOUNT = 'form';
+
+const IS_PROD = process.env.NODE_ENV === 'production';
```
Comment (zackharley) [inline client/ClientStore.js]: We check this in multiple places. Should we make an interface for environment variables?

File: server/api/v1/routes/user.routes.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+const USER = 'users';
+
+module.exports = (ctr) => {
+	const userAPI = Router();
+
+	const { user } = ctr;
+
+	userAPI.get('/users', (req, res) => {
+		user.getAllUsers().then((users) => {
```
Comment (zackharley) [inline server/api/v1/routes/user.routes.js]: This is essentially controller logic. I think our `*.routes.js` files should only include route declarations and middleware assignment:
```js
req.get('/users', usersCtrl.getAllUsers);
```

File: server/auth/auth.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		type: "AUTHORIZATION"
+	}
+};
+
+const ERROR_MESSAGES = {
+	MISSING_TOKEN: "Missing auth token!",
+	INVALID_TOKEN: "Invalid token!"
+};
+
+function createError(errorTemplate, message, data = {}) {
```
Comment (zackharley) [inline server/auth/auth.js]: Should make this a utility function since it's user everywhere

File: server/ctrs/index.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const authCtr = require('./auth.controller.js');
+const userCtr = require('./user.controller.js');
+
+module.exports = (db) => {
```
Comment (zackharley) [inline server/ctrs/index.js]: We can probably stop passing around `db` now

File: server/ctrs/user.controller.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	return Object.assign({}, errorTemplate, { message, data });
+}
+
+module.exports = (db) => {
+	const userCtr = {};
+
+	const { User } = require('../models');
+
+	userCtr.getUser = (userId) => new Promise((resolve, reject) => {
+		User.findOne({ _id: userId }).then((user) => {
```
Comment (zackharley) [inline server/ctrs/user.controller.js]: `User.findOne()` returns a Promise. There is no need to use the `new Promise` constructor. This goes for most of our controller logic.

File: server/db/db.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+module.exports = () => {
+    const host = process.env.MONGO_HOST;
+    const name = process.env.MONGO_DBNAME;
+    const port = process.env.MONGO_PORT || 27017;
+    const authSource = process.env.MONGO_AUTH_SOURCE;
+
+    if (!host || !name) {
+        winston.info("Missing Mongo enviroment variable MONGO_HOST or MONGO_DBNAME!");
```
Comment (zackharley) [inline server/db/db.js]: We should create a custom `winston` logger and use `winston.error()`

File: server/db/db.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        throw new Error();
+    }
+
+    const user = process.env.MONGO_USER;
+    const pass = process.env.MONGO_PASS;
+    const baseConnectionOptions = { useMongoClient: true };
+
+    if (!user || !pass) winston.info("Running Mongo unauthenticated!");
```
Comment (zackharley) [inline server/db/db.js]: Should be at least a `notice` log, if not `warn`

File: server/index.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+const path = require('path');
+const errorReporting = require('@google-cloud/error-reporting')();
+
+const isProd = process.env.NODE_ENV === 'production';
+
+(isProd)
+	? winston.info("Running production build!")
+	: winston.info("Running development build!");
+
+const auth = require('./auth/auth');
```
Comment (zackharley) [inline server/index.js]: Use `./auth/index.js` instead

File: server/index.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+const api = require('./api/api');
```
Comment (zackharley) [inline server/index.js]: Use `./api/index.js` instead

File: server/index.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+let ctr = require('./ctrs');
+const db = require('./db/db')();
```
Comment (zackharley) [inline server/index.js]: Use `./db/index.js` instead

File: server/index.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+const app = express();
+const port = process.env.PORT || 3000;
+
+// Third Party Middleware
+app.use(compression());
+app.use(bodyParser.json({ limit: '50mb' }));
+app.use(bodyParser.urlencoded({ extended: true }));
+
+db((err, db) => {
```
Comment (zackharley) [inline server/index.js]: Can remove this complicated logic as the db no longer needs to be passed around

File: server/mailer/index.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+const { SENDGRID_API_KEY, EMAIL_URL_HOST } = process.env;
+
+sendgrid.setApiKey(SENDGRID_API_KEY);
+
+const QHACKS_NO_REPLY_EMAIL = 'no-reply@qhacks.io';
+
+const APPLICATION_SUCCESSFUL = 'application-successful';
+
+const EMAIL_TEMPLATES = {
+    [APPLICATION_SUCCESSFUL]: fs.readFileSync(path.resolve(__dirname, `../../email/dist/${APPLICATION_SUCCESSFUL}.html`)).toString()
```
Comment (zackharley) [inline server/mailer/index.js]: Should create a way to generalize this line for all emails

File: server/models/userModel.js [hunk 13]
```diff
@@ -0,0 +1,120 @@
+const mongoose = require('mongoose');
+const bcrypt = require('bcrypt');
+const uuid = require('uuid');
+const mlhSchools = require('../../client/components/utils/mlhSchools.json');
```
Comment (zackharley) [inline server/models/userModel.js]: Need to bring this file serverside and server it up during client bootstrapping

File: server/models/userModel.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        type: String
+    },
+    refreshToken: {
+        type: String
+    }
+});
+
+UserSchema.static({
+    authenticate(email, password) {
+        return new Promise((resolve, reject) => {
```
Comment (zackharley) [inline server/models/userModel.js]: Don't need Promise constructor

File: server/models/userModel.js [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+                        });
+                });
+        });
+    }
+});
+
+UserSchema.method({
+    validPassword(password) {
+        return new Promise((resolve, reject) => {
+            bcrypt.compare(password, this.password, (err, valid) => {
```
Comment (zackharley) [inline server/models/userModel.js]: Pretty sure `bcrypt` can return a Promise so no need for the Promise constructor in that case

File: server/models/userModel.js [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * Helper method to hash password before saving.
+ * @param {Function} next Move to next middleware.
+ * @return {Function} Next middleware.
+ */
+UserSchema.pre('save', function(next) {
+    if (!this.isModified('password')) return next();
+
+    bcrypt.genSalt(SALT_WORK_FACTOR, (err, salt) => {
```
Comment (zackharley) [inline server/models/userModel.js]: This should not be using callbacks

File: server/models/userModel.js [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+        if (err) return next(err);
+
+        bcrypt.hash(this.password, salt, (err, hash) => {
```
Comment (zackharley) [inline server/models/userModel.js]: This should not be using callbacks

File: server/services/custom-validator.js [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+const ERROR_MESSAGES = {
+	INVALID_DATE: "The date provided is invalid! Be sure to follow ISO or RFC2822 date format.",
+	INVALID_EMAIL: "The email address provided is invalid!",
+	INVALID_GRAD_YEAR: "The graduation year provided is invalid!",
+	INVALID_PASSWORD: "The password provided is invalid, it must be more than eight characters.",
+	INVALID_NAME: "The first and last name combination provided is invalid.",
+	INVALID_PHONE_NUMBER: "The phone number provided is invalid."
+};
+
+class ValidationError extends Error {
```
Comment (zackharley) [inline server/services/custom-validator.js]: Let's get rid of this and just use the base error class

File: server/strings/index.js [hunk 19]
```diff
@@ -0,0 +1,7 @@
+const MAILER = require('./mailer.strings');
```
Comment (zackharley) [inline server/strings/index.js]: Maybe move strings files to their component folder? i.e. `mailer/mailer.strings.js`?

File: server/webhook/webhook.config.js [hunk 20]
```diff
@@ -0,0 +1,98 @@
+const moment = require('moment');
+
+const { NODE_ENV } = process.env;
+
+const config = {
+    '/api/v1/auth/signup': (user) => _toSlackMessage(user)
+};
+
+function _toSlackMessage(user) {
+    const fields = [
```
Comment (zackharley) [inline server/webhook/webhook.config.js]: Can this be generalized and made more compact, while still being explicit?

File: server/webhook/webhook.js [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+const webhookConfig = require('./webhook.config');
+const validator = require('validator');
+
+const { SLACK_WEBHOOK_URL } = process.env;
+
+if (!_isValidUrl(SLACK_WEBHOOK_URL)) {
+	winston.log('warn', 'No valid (or invalid) Slack webhook URL supplied');
+}
+
+function _createWebhookMiddleware(config) {
```
Comment (zackharley) [inline server/webhook/webhook.js]: Generalize this middleware for any webhook we want

File: .github/ISSUE_TEMPLATE.md [hunk 22]
```diff
@@ -0,0 +1,8 @@
+**Problem**
+
+**Requirements**
```
Comment (RobertWSaunders) [inline .github/ISSUE_TEMPLATE.md]: we should make these consistent between all of our enviroments

File: client/ClientStore.js (hunk 1 above)
Comment (RobertWSaunders) [reply client/ClientStore.js]: for file context I think its ok, also the check is a one liner

File: client/HackerStore/HackerActions.js [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+	"INVOKE_API_FAIL": "@@/hackerMiddleware/INVOKE_API_FAIL"
+};
+
+/**
+ * API Middleware request action types.
+ * NOTE: This action type will be dispatched from middleware
+ * upon validation from middleware, use to update status flags in reducer.
+ * @type {Object}
+ */
+const apiRequestTypes = {
```
Comment (RobertWSaunders) [inline client/HackerStore/HackerActions.js]: I want to clean these actions up to be more descriptive etc and make sure they are in sync with server language

File: client/HackerStore/HackerReducer.js [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+
+	isPasswordUpdated: false,
+	isPasswordUpdatedError: false,
+	isPasswordUpdatedLoading: false,
+
+	loginLoading: false,
+	loginError: false,
+	applicationError: false,
+	applicationLoading: false,
+    applicationFormErrorMessages: [],
```
Comment (RobertWSaunders) [inline client/HackerStore/HackerReducer.js]: need to put error state in a separate place

File: client/assets/semantic-ui/site/elements/button.overrides [hunk 25]
```diff
@@ -0,0 +1,21 @@
+//.ui.primary.button {
```
[the remark this answers is not in the document]
Comment (RobertWSaunders) [reply client/assets/semantic-ui/site/elements/button.overrides]: agreed

File: client/components/Dashboard/Dashboard.js [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+				<MenuBar onLogoutClick={this.handleLogoutClick.bind(this)} />
+
+				<Container text style={{ marginTop: '3em' }}>
+					<Segment raised>
+						<Header as='h2'><Emoji text="Congratulations, your application is in! :tada: :heart:" /></Header>
+						<p>Your application to attend QHacks 2018 has been submitted! So, what are the next steps? </p>
+						<p> We will be accepting applications until January 8th 2018. Later that week we will be sending emails to all of the applicants updating them on their application status.</p>
+						<p>If you are accepted, your next step will be to RSVP right here in the dashboard. You must complete this step within one week of receiving your acceptance email. The RSVP will be used to confirm your attendance to QHacks 2018.</p>
+						<p>What if I don't get accepted? There's a chance you may be added to the waitlist. If someone who has been accepted fails to RSVP or can't come to the event, then we will fill their spot with someone from the waitlist.</p>
+						<p>While you wait to hear from us there are some things you can do, such as:</p>
```
Comment (RobertWSaunders) [inline client/components/Dashboard/Dashboard.js]: man I hate this glitch so much

File: server/api/v1/routes/auth.routes.js [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+	// get auth controller
+	const { auth } = ctr;
+
+	authAPI.use(`/${AUTH}/session`, (req, res) => {
+		const { email, password } = req.body;
+
+		auth.authenticateUser(email, password).then((userInfo) => {
+			res.status(200).json(userInfo);
+		}).catch((err) => {
+			res.status(err.code).json(err);
```
[the remark this answers is not in the document]
Comment (RobertWSaunders) [reply server/api/v1/routes/auth.routes.js]: yeah agreed

File: server/api/v1/routes/user.routes.js (hunk 2 above)
Comment (RobertWSaunders) [reply server/api/v1/routes/user.routes.js]: I think its important to have the response happen where the request first lands, makes it easy to see whats happening, thoughts?

File: server/auth/auth.js (hunk 3 above)
Comment (RobertWSaunders) [reply server/auth/auth.js]: yep

File: server/ctrs/auth.controller.js [hunk 28]
```diff
@@ -0,0 +1,222 @@
...
+const _ = require('lodash');
+
+const JWT_ISSUER = 'QHacks Authentication';
+const ACCESS_TOKEN_EXPIRE_TIME = '5 minutes';
+const REFRESH_TOKEN_EXPIRE_TIME = '60 minutes';
+const QHACKS_2018_SLUG = 'qhacks-2018';
+
+const { AUTH_SECRET } = process.env;
+
+const ERRORS = {
```
Comment (RobertWSaunders) [inline server/ctrs/auth.controller.js]: I want to put these in the strings logic you made for mailers

File: server/ctrs/user.controller.js (hunk 5 above)
Comment (RobertWSaunders) [reply server/ctrs/user.controller.js]: The benefit of having the `new Promise` is that we can decide what we want to send back to the caller and not have a third party dictate it for us, in some scenarios this is needed when we add the refresh token etc. to the logic

File: server/db/db.js (hunk 6 above)
Comment (RobertWSaunders) [reply server/db/db.js]: yup, i also like pino-http for req logging, but yeah

File: server/index.js (hunk 8 above)
Comment (RobertWSaunders) [reply server/index.js]: im indifferent on this and won't change but if you want to yeah

File: server/index.js (hunk 9 above)
Comment (RobertWSaunders) [reply server/index.js]: im indifferent on this and won't change but if you want to yeah

File: server/index.js (hunk 10 above)
Comment (RobertWSaunders) [reply server/index.js]: im indifferent on this and won't change but if you want to yeah

File: server/index.js (hunk 11 above)
Comment (RobertWSaunders) [reply server/index.js]: keep the logic just dont pass into controller, I wouldnt say its complicated though

File: server/models/userModel.js (hunk 13 above)
Comment (RobertWSaunders) [reply server/models/userModel.js]: Yup

File: server/models/userModel.js (hunk 14 above)
Comment (RobertWSaunders) [reply server/models/userModel.js]: why not?

File: server/models/userModel.js (hunk 15 above)
Comment (RobertWSaunders) [reply server/models/userModel.js]: I couldn't find that it could

File: server/models/userModel.js (hunk 16 above)
Comment (RobertWSaunders) [reply server/models/userModel.js]: see last comment

File: server/models/userModel.js (hunk 17 above)
Comment (RobertWSaunders) [reply server/models/userModel.js]: see previous comment

File: server/services/custom-validator.js (hunk 18 above)
Comment (RobertWSaunders) [reply server/services/custom-validator.js]: agreed

File: server/strings/index.js (hunk 19 above)
Comment (RobertWSaunders) [reply server/strings/index.js]: if we implement strings logic for error we can keep it in the `strings` directory

**Pull request merged**